Pull or request changes

stefanjwojcik opened the pull request:
Hi Randy! 

I tried a lot of different things in order to provide cursoring functionality to as many functions as possible. I really wanted a macro approach, but that ultimately seemed overly complicated and not easy to debug. I still like the idea for some things (e.g. API limit monitoring), but it doesn't seem appropriate for cursoring. 

What I ended up doing was creating two cursoring methods - one for functions that give back Dict objects and another for functions that give back Arrays. 

**Cursoring methods for functions that give back Dicts:**
These are the more normal cursoring functions that actually give back a cursor object. These are the functions  like 'get_followers_ids' that return a Dict object where one of the objects is an array of "ids". The function I wrote simply grabs the 'next_cursor' and feeds it into the next call. It keeps calling until it reaches the 'count' desired in the call. It changes nothing about the original function call. 

**Cursoring methods for functions that give back Arrays:**
These generally give back Arrays of tweets like 'get_user_timeline' or 'get_home_timeline'. Cursoring over these objects is done, as you know, by supplying the 'max_id' option in subsequent calls. The API sends tweets in reverse order starting from the 'max_id', so the minimum tweet is always the oldest. 

**Cursoring methods for search_tweets:**
This function gives back a Dict object, but acts like the Array functions in that it gives back tweets with a 'max_id' argument. Cursoring works fine for this function, but the API doesn't seem to like to give back the correct number of desired records. I've found that you can set the number very high and you'll get more records, but you don't get as many as you ask for. It seems like an API issue, but I will continue to stew on it. 

**@twitter macro:**
I think it would be cool to work on a macro that monitors API usage so you don't have to keep track of it yourself. I think this will be fairly simple to build, but doesn't seem to belong in this pull. 

Happy to make changes to anything, this has been a fun exercise for me.

randyzwitch commented:
Thanks! I’ll take a look this week when I’m back on a computer 

> On Jan 19, 2020, at 4:35 PM, Stefan Wojcik <notifications@github.com> wrote:
> 
> Hi Randy!
> 
> I tried a lot of different things in order to provide cursoring functionality to as many functions as possible. I really wanted a macro approach, but that ultimately seemed overly complicated and not easy to debug. I still like the idea for some things (e.g. API limit monitoring), but it doesn't seem appropriate for cursoring.
> 
> What I ended up doing was creating two cursoring methods - one for functions that give back Dict objects and another for functions that give back Arrays.
> 
> Cursoring methods for functions that give back Dicts:
> These are the more normal cursoring functions that actually give back a cursor object. These are the functions like 'get_followers_ids' that return a Dict object where one of the objects is an array of "ids". The function I wrote simply grabs the 'next_cursor' and feeds it into the next call. It keeps calling until it reaches the 'count' desired in the call. It changes nothing about the original function call.
> 
> Cursoring methods for functions that give back Arrays:
> These generally give back Arrays of tweets like 'get_user_timeline' or 'get_home_timeline'. Cursoring over these objects is done, as you know, by supplying the 'max_id' option in subsequent calls. The API sends tweets in reverse order starting from the 'max_id', so the minimum tweet is always the oldest.
> 
> Cursoring methods for search_tweets:
> This function gives back a Dict object, but acts like the Array functions in that it gives back tweets with a 'max_id' argument. Cursoring works fine for this function, but the API doesn't seem to like to give back the correct number of desired records. I've found that you can set the number very high and you'll get more records, but you don't get as many as you ask for. It seems like an API issue, but I will continue to stew on it.
> 
> @twitter macro:
> I think it would be cool to work on a macro that monitors API usage so you don't have to keep track of it yourself. I think this will be fairly simple to build, but doesn't seem to belong in this pull.
> 
> Happy to make changes to anything, this has been a fun exercise for me.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/randyzwitch/Twitter.jl/pull/33
> 
> Commit Summary
> 
> added functionality for cursoring
> play file
> Add files via upload
> Update sandbox.jl
> deleting sandbox
> adding cursoring functions and docstrings
> Merge pull request #1 from stefanjwojcik/wojcik_cursorize
> blank out old functions
> changed order of imports
> fixed typo in home timeline endp
> fixed a bug where data array was empty
> fixed inf loop in cursorable
> still squashing timeline bug
> fixed overwite of get_search_tweets
> added some tests to runtests
> fixing max_id bug so it is always updated when since_id is provided
> missing parens
> options object parsing incorrectly
> more bug squash
> tests of cursoring arrays
> eliminate unnecessary branch in cursoring
> creating an inner loop within cursoring to parse options for search results
> missing end
> proper cursoring in search
> updated tests (passed) and added more docs to cursoring functions
> File Changes
> 
> M src/Twitter.jl (1)
> A src/cursoring.jl (372)
> M src/macrogenerated.jl (14)
> M test/runtests.jl (36)
> Patch Links:
> 
> https://github.com/randyzwitch/Twitter.jl/pull/33.patch
> https://github.com/randyzwitch/Twitter.jl/pull/33.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

randyzwitch commented on test/runtests.jl:
```diff
@@ -1,7 +1,11 @@
 
 twitterauth(ENV["CONSUMER_KEY"], ENV["CONSUMER_SECRET"], ENV["ACCESS_TOKEN"], ENV["ACCESS_TOKEN_SECRET"])
 
+TWITTERCRED = twitterauth(ENV["CONSUMER_KEY"], ENV["CONSUMER_SECRET"], ENV["ACCESS_TOKEN"], ENV["ACCESS_TOKEN_SECRET"])
```
For this PR, this line is actually unneeded. In the bigger picture, we might consider not having `twitterauth` return globally, since people could might want to pull using multiple API keys for some reason. Though, having to pass a credential object to each function is also annoying

stefanjwojcik replied:
Yes, good catch. I had included this line for some line-by-line testing.

randyzwitch commented on test/runtests.jl:
```diff
@@ -1,7 +1,11 @@
+
+
 #get_mentions_timeline
 mentions_timeline_default = get_mentions_timeline()
```
This test fails with

```julia
julia> mentions_timeline_default = get_mentions_timeline()
ERROR: KeyError: key "count" not found
Stacktrace:
 [1] getindex at ./dict.jl:477 [inlined]
 [2] parse_results(::Bool, ::Array{Any,1}, ::Dict{String,Any}, ::Array{Any,1}, ::Int64) at /home/rzwitch/.julia/dev/Twitter/src/cursoring.jl:76
 [3] cursor(::Bool, ::Array{Any,1}, ::Dict{String,Any}, ::String, ::Int64) at /home/rzwitch/.julia/dev/Twitter/src/cursoring.jl:139
 [4] #get_mentions_timeline#7(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(get_mentions_timeline)) at /home/rzwitch/.julia/dev/Twitter/src/cursoring.jl:244
 [5] get_mentions_timeline() at /home/rzwitch/.julia/dev/Twitter/src/cursoring.jl:230
 [6] top-level scope at none:0
```

randyzwitch replied:
Looks like some of the methods need a `count::Int = 20` default kwarg, `get_home_timeline()` fails as well

```julia
julia> home_timeline_default = get_home_timeline()
ERROR: KeyError: key "count" not found
Stacktrace:
 [1] getindex at ./dict.jl:477 [inlined]
 [2] parse_results(::Bool, ::Array{Any,1}, ::Dict{String,Any}, ::Array{Any,1}, ::Int64) at /home/rzwitch/.julia/dev/Twitter/src/cursoring.jl:76
 [3] cursor(::Bool, ::Array{Any,1}, ::Dict{String,Any}, ::String, ::Int64) at /home/rzwitch/.julia/dev/Twitter/src/cursoring.jl:139
 [4] #get_home_timeline#9(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(get_home_timeline)) at /home/rzwitch/.julia/dev/Twitter/src/cursoring.jl:307
 [5] get_home_timeline() at /home/rzwitch/.julia/dev/Twitter/src/cursoring.jl:293
 [6] top-level scope at none:0
```

stefanjwojcik replied:
added default count = 1. Tests pass.

stefanjwojcik commented:
Alright, fixed the default count error - now runs fine. I also fixed a type error in runtests.jl that was throwing an error in testing - arrays weren't coming out at Arrays of Tweets. That is fixed too. Lastly, there were two incorrect variable refs (undefined) - fixed those as well. All tests pass.

randyzwitch commented:
I think this is everything, thanks! Going to squash/merge, and then after it passes CI, I'll push a new version
